### ISSUE TITLE
Last dimension of the npy feature map data is different from csv files

### ISSUE THREAD

**Issue by MusenZR** [SizheAn/MARS#3]
As discribed in the paper, the last dimension of feature map is suppose to be reflection intensity, the csv and *.mat files indicates that the fifth d data is a int number. But after loading the numpy feature map, it can be seen that the last d is a negative float number.I parsed the csv data into pytorch dataloader and run it in pytorch env, MSE loss show a difference from running the zipped numpy feature map.

from the numpy feature map
[[-0.45508     0.96094    -1.7871      0.71201    -0.53580797]
  [-0.17871     1.6396      0.95117     0.356      -0.61807922]
  [ 0.          2.0469     -0.61914     0.71201    -0.55637578]
  [ 0.12988     2.0742     -0.055664    1.068      -0.53580797]
  [ 0.18652     1.9561      0.32324    -0.356      -0.28899424]
  [ 0.18945     1.9736      0.39355     0.71201    -0.26842643]
  [ 0.18945     1.9883      0.30664     0.356       0.80109974]
  [ 0.18945     1.9961      0.25195    -0.356      -0.12445175]]
same data from csv files:
[[-4.5508e-01  9.6094e-01 -1.7871e+00  7.1201e-01  9.0000e+00]
  [-1.7871e-01  1.6396e+00  9.5117e-01  3.5600e-01  5.0000e+00]
  [ 0.0000e+00  2.0469e+00 -6.1914e-01  7.1201e-01  8.0000e+00]
  [ 1.2988e-01  2.0742e+00 -5.5664e-02  1.0680e+00  9.0000e+00]
  [ 1.8652e-01  1.9561e+00  3.2324e-01 -3.5600e-01  2.1000e+01]
  [ 1.8945e-01  1.9736e+00  3.9355e-01  7.1201e-01  2.2000e+01]
  [ 1.8945e-01  1.9883e+00  3.0664e-01  3.5600e-01  7.4000e+01]
  [ 1.8945e-01  1.9961e+00  2.5195e-01 -3.5600e-01  2.9000e+01]]

You can see the difference in the last dim.
So what's the last dim of the numpy feature map? Did you do some calculation to optimize the raw intensity?

**Comment by SizheAn** [comment]
Hi,

Thanks for the interest in the work. I normalize the power intensity into [-1,1] range when preprocessed the data as follow:

`
normed_intensity = (intensity - mu)/sigma
`

**Comment by MusenZR** [comment]
Thanks for the quick response! 
I've got another question though. I've run the pytorch version of your model, and also tested your keras version.  But there's a gap between train loss and the val loss in the first epoch, as in the picture under.
![image](https://user-images.githubusercontent.com/65099446/189534957-fe78df5a-1bcd-4fb3-8e28-ab39c3be65a4.png)
They eventually come close, but is there a explaination of this gap at the beginning of training?

**Comment by SizheAn** [comment]
> Thanks for the quick response! I've got another question though. I've run the pytorch version of your model, and also tested your keras version. But there's a gap between train loss and the val loss in the first epoch, as in the picture under. ![image](https://user-images.githubusercontent.com/65099446/189534957-fe78df5a-1bcd-4fb3-8e28-ab39c3be65a4.png) They eventually come close, but is there a explaination of this gap at the beginning of training?

Emm not sure the exact reason. Could be weight initialization, or the optimizer's hyper parameters of different framework. I designed the model first on Keras then switched to Pytorch later. Since the loss and MSE converged to a similar value I did not do a deep dive. But I strongly suspect that it could be some parameters discrepancy between two frameworks. Let me know if you find out the reason :)

**Comment by MusenZR** [comment]
Hi there! I examined my training code and find out that the 'scheduler.step()' has been misplaced in the 'with torch.no_grad():' part. I moved it to the end of epoch loop code and the loss curve are basically the same now.
![image](https://user-images.githubusercontent.com/65099446/189574079-0be2e7ab-470b-4b93-8093-1b3239c2c7e1.png)

**Comment by MusenZR** [comment]
Hi, there. I've got another data problem.
I processed your csv files on my own, and saved them to *.npy files. But the mse loss of my processed data are a little bit higher than your github original *.npy data. Your data can reach a 0.0065 mse loss on validation set, but my processed data can only reach about 0.0074. I've tested multiple times and the results are the same.
Here are my codes to process the csv files.
```
def process_radar_data(self, path):
        holder = np.zeros((1, 8, 8, 5))
        for i in path:
            feature = []
            with open(i, encoding='utf-8') as f:
                data = np.loadtxt(f, dtype=float, delimiter=",", skiprows=1, usecols=(0, 2, 3, 4, 5, 6))
            data[:, 5] = (data[:, 5] - data[:, 5].mean()) / data[:, 5].std()
            data = np.split(data[:, 1:], np.unique(data[:, 0], return_index=True)[1][1:])
            for frame in data:
                frame = np.array(sorted(map(tuple, frame)))
                if frame.shape[0] < 64:
                    z = np.zeros((64 - frame.shape[0], 5))
                    frame = np.vstack((frame, z))
                else:
                    index = np.random.choice(frame.shape[0], size=64, replace=False)
                    frame = frame[index]
                frame = frame.reshape((8, 8, 5))
                feature.append(frame)
            feature = np.array(feature)
            holder = np.concatenate((holder, feature))
        self.radar_data = holder[1:]
```

```
    def process_kinect_data(self, path, joints):
        j = joints
        holder = np.zeros((1, j * 3))
        for i in path:
            with open(i, encoding='utf-8') as f:
                data = np.loadtxt(f, dtype=float, delimiter=",", skiprows=1)
                label = data
            if j == 19:
                index = [7, 11, 21, 22, 23, 24,
                         32, 36, 46, 47, 48, 49,
                         57, 61, 71, 72, 73, 74]
                label = np.delete(data, index, axis=1)
            holder = np.concatenate((holder, label))
        self.labels_data = holder[1:]
```

**Comment by Carbord** [comment]
> > Thanks for the quick response! I've got another question though. I've run the pytorch version of your model, and also tested your keras version. But there's a gap between train loss and the val loss in the first epoch, as in the picture under. ![image](https://user-images.githubusercontent.com/65099446/189534957-fe78df5a-1bcd-4fb3-8e28-ab39c3be65a4.png) They eventually come close, but is there a explaination of this gap at the beginning of training?
> 
> Emm not sure the exact reason. Could be weight initialization, or the optimizer's hyper parameters of different framework. I designed the model first on Keras then switched to Pytorch later. Since the loss and MSE converged to a similar value I did not do a deep dive. But I strongly suspect that it could be some parameters discrepancy between two frameworks. Let me know if you find out the reason :)

> Thanks for the quick response! I've got another question though. I've run the pytorch version of your model, and also tested your keras version. But there's a gap between train loss and the val loss in the first epoch, as in the picture under. ![image](https://user-images.githubusercontent.com/65099446/189534957-fe78df5a-1bcd-4fb3-8e28-ab39c3be65a4.png) They eventually come close, but is there a explaination of this gap at the beginning of training?

Hello! I’m also trying to reproduce the code in pytorch. But my mse loss of validation set is about 1 at 150 epochs. Will you share your pytorch version code? Or would you be kind to help checking the following reproduced codes? Thanks a lot!
![train](https://user-images.githubusercontent.com/33995512/221805649-33ddccee-bdac-4f90-8a0c-78cf694b46c3.png)
![model](https://user-images.githubusercontent.com/33995512/221805688-e672a03f-f26c-4d29-83d9-5f16fb2eeea5.png)
